network/src/transport/mod.rs

```diff
@@ -176,14 +176,14 @@ pub struct Connection<TSocket> {
 }
 
 /// Convenience function for adding a timeout to a Future that returns an `io::Result`.
-async fn timeout_io<F, T>(duration: Duration, fut: F) -> io::Result<T>
+async fn timeout_io<F, T>(time_service: TimeService, duration: Duration, fut: F) -> io::Result<T>
 where
     F: Future<Output = io::Result<T>>,
 {
-    let res = timeout(duration, fut).await;
+    let res = time_service.timeout(duration, fut).await;
     match res {
         Ok(out) => out,
-        Err(err) => Err(io::Error::new(io::ErrorKind::TimedOut, err)),
+        Err(Elapsed) => Err(io::Error::new(io::ErrorKind::TimedOut, Elapsed)),
```

Comment:
since it's using the time-service timeout Elapsed struct, we can actually just match on the struct here. tokio was doing #[non_exhaustive_match] for API stability reasons, so you couldn't do this before...

this way we'll get a compiler error if the error cases change for the timeout future